### PR TITLE
fix: always include self-referential transclusion

### PR DESCRIPTION
**Pull request by linonetwo**
This will ignore current tiddler when collecting transclusions.

But this is different from the behavior of backlinks[] !

---

Update: restore old behavior: include itself when there is a self-referential transclusion, so it works the same as `backlinks[]`

### CODE REVIEW

**Comment by kookma** [comment]
> This will ignore current tiddler when collecting transclusions.
> 
> But this is different from the behavior of backlinks[] !

I would recommend seeing the same behavior for all node filter operator! Different behavior makes it difficult to remember.

**Comment by linonetwo** [comment]
@kookma Have a try if you have time, I add more test so it have a high probablity that it works as expected.

**Comment by kookma** [comment]
I did some tests and all previous issues have gone!

The only odd behavior is both `transcludes` and `backtranscludes` retune the currentTiddler as a node!
This is meaningless and confusing! So, I would recommend to solve this issue in the code!

That means if you create an empty tiddler with below scripts

```
<<list-links "[all[current]transcludes[]]">>

<<list-links "[all[current]backtranscludes[]]">>
```
Tiddlywiki will return

* New Tiddler
* New Tiddler

**Comment by linonetwo** [comment]
The behavior of links and backlinks is similar, but I will check it later, after I finish my burgerking.

**Comment by kookma** [comment]
> The behavior of links and backlinks is similar, but I will check it later, after I finish my burgerking.

I do not think so! For example, create a new tiddler with below script:

```
{{{ [all[current]links[]] }}}
```

It returns nothing!

**Comment by pmario** [comment]
@kookma -- You are right - If this code is the only thing in a tiddler -> It should create empty lists. 

```
<<list-links "[all[current]transcludes[]]">>

<<list-links "[all[current]backtranscludes[]]">>
```

**Comment by linonetwo** [comment]
Oh, why `{{{ [all[current]backtranscludes[]] }}}` won't output tiddler title, but `<<list-links` will...

**Comment by linonetwo** [comment]
Added detailed rules to handle self-referencial cases.

Also outof professional ethics, I refactor links[] transcludes[] related wiki method to a file, so wiki.js won't be too long to edit.

File: core/modules/wiki-extract.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            value = parseTreeNode.attributes.tiddler.value;
+          } else if(parseTreeNode.attributes.$field && parseTreeNode.attributes.$field.type === "string") {
+            // Empty value (like `<$transclude $field='created'/>`) means self-referential transclusion. 
+            value = title;
+          } else if(parseTreeNode.attributes.field && parseTreeNode.attributes.field.type === "string") {
+            // Old usage with Empty value (like `<$transclude field='created'/>`)
+            value = title;
+          }
+          // Deduplicate the result.
+          if(value && transcludes.indexOf(value) === -1) {
```
Comment (Jermolene) [inline core/modules/wiki-extract.js:105]: I think you could use $tw.utils.pushTop() here for consistency.

**Review by Jermolene — COMMENTED** [review]
Thank you @linonetwo 

> Also outof professional ethics, I refactor links[] transcludes[] related wiki method to a file, so wiki.js won't be too long to edit. @Jermolene

I do agree that wiki.js is unwieldy at the moment, but refactoring like that should go into a separate PR. Otherwise it is very hard when reading the commit history to understand precisely what is changed in this fix.

In such a refactoring we might split wiki.js up into 5-10 separate files, and take a little time to get the naming right.

Another issue with this PR is that the whitespace is inconsistently using spaces vs. tabs. Please could you correct them to use tabs.

File: core/modules/wiki-extract.js (hunk 1 above)
Comment (linonetwo) [reply core/modules/wiki-extract.js]: Updated.

**Comment by linonetwo** [comment]
Maybe it worth some time to look at linter related PRs!

**Comment by kookma** [comment]
Thank you! All my previous tests are passed with latest release.

**Pull request merged**